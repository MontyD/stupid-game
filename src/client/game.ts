import { GameRunState } from '../models/entities/game';
import { ObjectOfAny } from '../utils/types';

export class Game {

    public static from(gameEntity: ObjectOfAny): Game {
        const game: ObjectOfAny = new this();
        const gameKeys = Object.keys(game);
        Object.keys(gameEntity).forEach(key => {
            if (gameKeys.includes(key)) {
                game[key]  = gameEntity[key];
            }
        });
        return game as Game;
    }

    public runState: GameRunState  = GameRunState.WAITING_FOR_PLAYERS_TO_JOIN;
    public code: string = '';
}
