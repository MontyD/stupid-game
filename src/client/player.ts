import { ObjectOfAny } from '../utils/types';
import { TypeOfPlayer } from '../models/entities/player';

export class Player {

    public static from(playerEntity: ObjectOfAny): Player {
        const player: ObjectOfAny = new Player();
        const playerKeys = Object.keys(player);
        Object.keys(playerEntity).forEach(key => {
            if (playerKeys.includes(key)) {
                player[key]  = playerEntity[key];
            }
        });
        return player as Player;
    }

    public id: string = '';
    public name: string = '';
    public type: TypeOfPlayer = TypeOfPlayer.OBSERVER;
    public isHost: boolean = false;
}
