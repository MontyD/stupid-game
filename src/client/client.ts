import { TopLevelClientToServerMessages, TopLevelServerToSingleClientMessages } from "../models/messages/top-level";
import { ObjectOfAny } from '../utils/types';
import { BroadcastGameMessages, SingleClientGameMessages, ClientToServerGameMessages } from '../models/messages/game';
import { JoinGameOptions } from "../controllers/game-controller";
import { PlayerDTO } from "../models/entities/player";
import { GameDTO, GameRunState } from "../models/entities/game";
import { GameDefinitionDTO } from "../models/entities/game-definition";
import { ValidationError } from "../controllers/validation-error";
import {
    BroadcastRoundMessages,
    ClientToServerRoundMessages,
    SingleClientRoundMessages
} from "../models/messages/round";
import { Prompt } from "../models/entities/prompt";

export type RequestCommands = TopLevelServerToSingleClientMessages |
                                BroadcastGameMessages |
                                SingleClientGameMessages |
                                BroadcastRoundMessages |
                                SingleClientRoundMessages;

export enum ClientEvent {
    PLAYER_JOINED,
    PLAYER_LEFT,
    GAME_STARTED,
    PLAYER_RESPONSE,
}

export class Client {

    public static createAsHost(socket: SocketIOClient.Socket): Promise<Client> {
        return (new Client(socket)).requestGameAsHost();
    }

    public static joinGameAsPlayer(socket: SocketIOClient.Socket, options: JoinGameOptions): Promise<Client> {
        return (new Client(socket)).joinGameAsPlayer(options);
    }

    public game?: GameDTO;
    public gameDefinition?: GameDefinitionDTO;
    public player?: PlayerDTO;
    public otherPlayers: PlayerDTO[] = [];

    private eventHandlers: Map<ClientEvent, Array<(arg: ObjectOfAny) => void>> = new Map();

    private constructor(private socket: SocketIOClient.Socket) { }

    public async startGame() {
        this.socket.emit(ClientToServerGameMessages.START);
        await this.takeNext(BroadcastGameMessages.STARTED);
        await this.takeNext(BroadcastRoundMessages.STARTED);
    }

    public async instructionsComplete(): Promise<Prompt | null> {
        this.socket.emit(ClientToServerRoundMessages.READY_TO_TAKE_PROMPT);
        if (!this.player || this.player.isHost) {
            return null;
        }

        const response = await this.takeNext(SingleClientRoundMessages.PROMPT);
        return response as Prompt;
    }

    public async sendImageResponse(data: ArrayBuffer) {
        this.socket.emit(ClientToServerRoundMessages.RESPONSE, { data });
        this.takeNext(SingleClientRoundMessages.RESPONSE_RECEIVED);
    }

    public async waitForNextPrompt() {
        this.takeNext(BroadcastRoundMessages.PROMPT_COMPLETE);
        this.socket.emit(ClientToServerRoundMessages.READY_TO_TAKE_PROMPT);
    }

    public on(event: ClientEvent, handler: (arg: ObjectOfAny) => void) {
        const handlers = this.eventHandlers.get(event) || [];
        handlers.push(handler);
        this.eventHandlers.set(event, handlers);
    }

    public off(event: ClientEvent, handler: (arg: ObjectOfAny) => void) {
        const handlers = (this.eventHandlers.get(event) || []).filter(
            existingHandler => existingHandler !== handler
        );
        this.eventHandlers.set(event, handlers);
    }

    public dispose(): void {
        this.socket.disconnect();
    }

    private async requestGameAsHost(): Promise<this> {
        this.socket.emit(TopLevelClientToServerMessages.CREATE_GAME);

        const { game, player } = await this.takeNext(BroadcastGameMessages.CREATED);
        this.game = game;
        this.player = player;
        this.setupHandlers();
        return this;
    }

    private async joinGameAsPlayer(options: JoinGameOptions): Promise<this> {
        this.socket.emit(TopLevelClientToServerMessages.JOIN_GAME, options);

        const {
             game, player, otherPlayers,
        } = await this.takeNext(SingleClientGameMessages.JOIN_SUCCESSFUL);
        this.game = game;
        this.player = player;
        this.otherPlayers = otherPlayers;
        this.setupHandlers();
        return this;
    }

    private handlePlayerJoined({ game, player }: { game: GameDTO, player: PlayerDTO }): void {
        this.game = game;
        this.otherPlayers.push(player);
        (this.eventHandlers.get(ClientEvent.PLAYER_JOINED) || []).forEach(handler => handler(player));
    }

    private handlePlayerLeft({ player: gonePlayer }: { player: PlayerDTO}): void {
        this.otherPlayers = this.otherPlayers.filter(existingPlayer => existingPlayer.id !== gonePlayer.id);
        (this.eventHandlers.get(ClientEvent.PLAYER_LEFT) || []).forEach(handler => handler(gonePlayer));
    }

    private handleGameStarted({ gameDefinition }: { gameDefinition: GameDefinitionDTO }): void {
        if (!this.game) {
            return;
        }
        this.game.runState = GameRunState.RUNNING_ROUND;
        this.gameDefinition = gameDefinition;
        (this.eventHandlers.get(ClientEvent.GAME_STARTED) || []).forEach(handler => handler(gameDefinition));
    }

    private handlePlayerResponse(res: { playerId: string }): void {
        (this.eventHandlers.get(ClientEvent.PLAYER_RESPONSE) || []).forEach(handler => handler(res));
    }

    private async takeNext(command: RequestCommands): Promise<ObjectOfAny> {
        const result = await Promise.race([
            this.handlerToKeyedPromise<ObjectOfAny>('success', command),
            this.handlerToKeyedPromise<string>(
                'validationError',
                TopLevelServerToSingleClientMessages.VALIDATION_ERROR
            ),
            this.handlerToKeyedPromise<string>('failure', TopLevelServerToSingleClientMessages.ERROR),
        ]);

        if (result.validationError) {
            throw new ValidationError(result.validationError as string);
        }
        if (result.failure) {
            throw new Error(result.failure as string);
        }
        return result.success as ObjectOfAny;
    }

    private handlerToKeyedPromise<T>(keyName: string, eventName: string): Promise<{ [keyof: string]: T }> {
        return new Promise((resolve) => this.socket.once(eventName, (arg: any) => resolve({ [keyName]: arg })));
    }

    private setupHandlers() {
        this.socket.on(BroadcastGameMessages.PLAYER_JOINED, this.handlePlayerJoined.bind(this));
        this.socket.on(BroadcastGameMessages.PLAYER_DISCONNECTED, this.handlePlayerLeft.bind(this));
        this.socket.on(BroadcastGameMessages.STARTED, this.handleGameStarted.bind(this));
        this.socket.on(BroadcastRoundMessages.PLAYER_RESPONSE, this.handlePlayerResponse.bind(this));
    }

}
