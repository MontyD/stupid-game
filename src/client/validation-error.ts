import { RequestCommands } from './client';

export class ValidationError extends Error {

    public command: RequestCommands;

    constructor(message: string, command: RequestCommands) {
        super(message);
        this.command = command;
    }

}
