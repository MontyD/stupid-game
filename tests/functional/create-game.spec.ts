import { createClientSocket } from "./util/create-connection";
import { Client } from '../../src/client/client';
import 'jest';

describe('connection handling', () => {

    let socket: SocketIOClient.Socket;
    let toDisconnect: SocketIOClient.Socket[] = [];

    beforeEach(() => {
        socket = createClientSocket();
        toDisconnect.push(socket);
    });

    afterEach(() => {
        toDisconnect.forEach(io => io.disconnect());
        toDisconnect = [];
    });

    it('will create a game', async () => {
        const {game, player, otherPlayers} = await Client.createAsHost(socket);
        expect(game!.code.length).toEqual(5);
        expect(game!.runState).toEqual('WAITING_FOR_PLAYERS_TO_JOIN');
        expect(player).not.toBeUndefined();
        expect(player!.name).toEqual('HOST');
        expect(player!.isHost).toBe(true);
        expect(otherPlayers).toHaveLength(0);
    });

    it('will allow players to join the game', async () => {
        const playerJoinedListener = jest.fn();
        const secondSocket = createClientSocket();
        toDisconnect.push(secondSocket);

        const host = await Client.createAsHost(socket);
        host.onPlayerJoined(playerJoinedListener);
        expect(host.otherPlayers).toEqual([]);

        const activeClient = await Client.joinGameAsPlayer(secondSocket, {
            playerName: 'new-player',
            gameCode: host.game!.code,
        });

        if (!activeClient.game || !host.game || !activeClient.player || !host.player) {
            throw new Error('game or player was null on activeClient or host');
        }

        expect(activeClient.game.id).toEqual(host.game.id);
        expect(activeClient.otherPlayers).toHaveLength(1);
        expect(activeClient.otherPlayers[0].id).toEqual(host.player.id);
        expect(activeClient.player.name).toEqual('new-player');

        expect(host.otherPlayers).toHaveLength(1);
        expect(host.otherPlayers[0]!.id).toEqual(activeClient.player.id);
        expect(host.otherPlayers[0]!.name).toEqual('new-player');

        expect(playerJoinedListener).toHaveBeenCalledWith(host.otherPlayers[0]);
    });

    it('will handle players disconnecting', async (done) => {
        const secondSocket = createClientSocket();
        const host = await Client.createAsHost(socket);
        const activePlayerClient = await Client.joinGameAsPlayer(secondSocket, {
            playerName: 'i-will-leave',
            gameCode: host.game!.code,
        });

        host.onPlayerLeft((player: any) => {
            expect(player.id).toEqual(activePlayerClient.player!.id);
            expect(host.otherPlayers).toEqual([]);
            done();
        });

        secondSocket.disconnect();
    });

    it('will limit the number of players to 12', async () => {
        const host = await Client.createAsHost(socket);
        const players = [];

        let i = 0;
        while (i++ < 12) {
            const newPlayerSocket = createClientSocket();
            const newPlayer = await Client.joinGameAsPlayer(newPlayerSocket, {
                playerName: `player ${i}`,
                gameCode: host.game!.code,
            });
            toDisconnect.push(newPlayerSocket);
            players.push(newPlayer);
        }

        try {
            const newPlayerSocket = createClientSocket();
            toDisconnect.push(newPlayerSocket);
            await Client.joinGameAsPlayer(newPlayerSocket, {
                playerName: 'one player too many',
                gameCode: host.game!.code,
            });
            // should never be here, should throw
            expect(true).toBe(false);
        } catch (exception) {
            expect(exception).toMatchSnapshot();
        }
    });

});
