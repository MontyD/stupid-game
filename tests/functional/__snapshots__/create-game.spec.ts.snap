// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`connection handling will limit the number of players to 12 1`] = `[ValidationError: Game is full]`;
